pkg/build/sbom.go

```diff
@@ -26,42 +26,38 @@ import (
 	"istio.io/release-builder/pkg/util"
 )
 
-// Sbom generates Software Bill Of Materials for istio repo in an SPDX readable format.
+const (
+	sbomOutputURI string = "https://storage.googleapis.com/istio-release/releases"
+)
+
+// GenerateBillOfMaterials generates Software Bill Of Materials for istio repo in an SPDX readable format.
 func GenerateBillOfMaterials(manifest model.Manifest) error {
 	// Retrieve istio repository path to run the sbom generator
 	istioRepoDir := manifest.RepoDir("istio")
-	sourceSbomFile := path.Join(manifest.OutDir(), "istio-source.spdx")
-	sourceSbomNamespace := fmt.Sprintf("https://storage.googleapis.com/istio-release/releases/%s/istio-source.spdx",
-		manifest.Version)
-	releaseSbomFile := path.Join(manifest.OutDir(), "istio-release.spdx")
-	releaseSbomNamespace := fmt.Sprintf("https://storage.googleapis.com/istio-release/releases/%s/istio-release.spdx",
-		manifest.Version)
-
-	// construct all the docker image tarball names as bom currently cannot accept directory as input
-	dockerDir := path.Join(manifest.OutDir(), "docker")
-	dockerImages := []string{}
-	if err := filepath.Walk(dockerDir, func(path string, fi os.FileInfo, err error) error {
-		if err != nil {
-			return err
-		}
-		if fi == nil {
-			return fmt.Errorf("failed to get fileinfo for file at path %s", path)
-		}
-		if fi.IsDir() {
-			return nil
-		}
-		dockerImages = append(dockerImages, path)
-		return nil
-	}); err != nil {
-		return fmt.Errorf("failed to walk directory %s: %v", dockerDir, err)
+	nameSpaceURI := sbomOutputURI
+	if manifest.BillOfMaterialsURI != "" {
+		nameSpaceURI = manifest.BillOfMaterialsURI
 	}
+	sourceSbomFile := path.Join(manifest.OutDir(), fmt.Sprintf("istio-source-%s.spdx", manifest.Version))
+	sourceSbomNamespace := path.Join(nameSpaceURI, manifest.Version, fmt.Sprintf("istio-source-%s.spdx", manifest.Version))
+
+	releaseSbomFile := path.Join(manifest.OutDir(), fmt.Sprintf("istio-source-%s.spdx", manifest.Version))
+	releaseSbomNamespace := path.Join(nameSpaceURI, manifest.Version, fmt.Sprintf("istio-source-%s.spdx", manifest.Version))
 
 	// Run bom generator to generate the software bill of materials(SBOM) for istio.
 	log.Infof("Generating Software Bill of Materials for istio release artifacts")
-	if err := util.VerboseCommand("bom", "--log-level", "error", "generate", "--name", "Istio Release "+manifest.Version,
-		"--namespace", releaseSbomNamespace, "--ignore", "licenses,'*.sha256',docker", "--dirs", manifest.OutDir(),
-		"--image-archive", strings.Join(dockerImages, ","), "--output", releaseSbomFile).Run(); err != nil {
-		return fmt.Errorf("couldn't generate sbom for istio release artifacts: %v", err)
+	// For Docker output in 'context' mode we will not produce SBOM.
+	// `bom` can produce bill only for tar and remote images.
+	if manifest.DockerOutput == model.DockerOutputTar {
```

Comment:
i am wondering where we are running SBOM in 'context' mode as the PR title says.

Reply:
Currently we don't call bom at all if context mode is specified. I am enabling it here https://github.com/istio/release-builder/pull/1430/files#diff-e3115b8ca9fdf611ae3dc61dc235e39aeeddbca61036b803ba3c0826f25cc3c9L91

For this snippet. I am skipping sbom generation for docker artifacts since `bom` doesn't support local context/registries but generate one for the source code.

Reply:
OK, so for source code we need to generate SBOM, as docker output mode does not matter in that case. Previously we were fully skipping. @puerco is this correct, that bom cannot support local registries? I remember you mentioning that it also should work.

Reply:
Could you please clarify how the PR title "[Run SBOM in 'context' mode]" makes sense now? You run it in context mode, but the SBOM is generated only for source code right?

Reply:
Updated the title to highlight that explicitly.